src/main/kotlin/uk/gov/justice/digital/hmpps/incidentreporting/jpa/Report.kt

```diff
@@ -297,63 +324,89 @@ class Report(
+    //       For Event's title/description may make less sense, but we're keeping this in
     //       as well for now.
-    incidentDateAndTime = upsert.incidentDateTime
-    event.eventDateAndTime = incidentDateAndTime
+    this.incidentDateAndTime = upsert.incidentDateTime
+    this.event.eventDateAndTime = incidentDateAndTime
 
-    location = upsert.prison.code
-    event.location = location
+    this.location = upsert.prison.code
+    this.event.location = location
 
-    title = upsert.title ?: NO_DETAILS_GIVEN
-    event.title = title
+    this.title = upsert.title ?: NO_DETAILS_GIVEN
+    this.event.title = title
 
-    description = upsert.description ?: NO_DETAILS_GIVEN
-    event.description = description
+    this.description = upsert.description ?: NO_DETAILS_GIVEN
+    this.event.description = description
 
-    reportedBy = upsert.reportingStaff.username
-    reportedAt = upsert.reportedDateTime
+    this.reportedBy = upsert.reportingStaff.username
+    this.reportedAt = upsert.reportedDateTime
 
     val newStatus = Status.fromNomisCode(upsert.status.code)
     if (newStatus != status) {
-      status = newStatus
+      this.status = newStatus
       addStatusHistory(newStatus, now, updatedBy)
     }
 
-    questionSetId = "${upsert.questionnaireId}"
-
-    createdAt = upsert.createDateTime
-    event.createdAt = upsert.createDateTime
-
-    modifiedAt = upsert.lastModifiedDateTime ?: upsert.createDateTime
-    event.modifiedAt = modifiedAt
+    this.questionSetId = "${upsert.questionnaireId}"
 
-    modifiedBy = updatedBy
-    event.modifiedBy = updatedBy
+    this.createdAt = upsert.createDateTime
+    this.event.createdAt = upsert.createDateTime
 
-    staffInvolved.clear()
-    addNomisStaffInvolvements(upsert.staffParties)
+    this.modifiedAt = upsert.lastModifiedDateTime ?: upsert.createDateTime
+    this.event.modifiedAt = modifiedAt
 
-    prisonersInvolved.clear()
-    addNomisPrisonerInvolvements(upsert.offenderParties)
+    this.modifiedBy = updatedBy
+    this.event.modifiedBy = updatedBy
 
-    correctionRequests.clear()
-    addNomisCorrectionRequests(upsert.requirements)
+    updateNomisStaffInvolvements(upsert.staffParties)
+    updateNomisPrisonerInvolvements(upsert.offenderParties)
+    updateNomisCorrectionRequests(upsert.requirements)
+    updateQuestionAndResponses(upsert.questions)
+    updateHistory(upsert.history)
+  }
 
-    questions.clear()
-    addNomisQuestions(upsert.questions)
+  fun updateQuestionAndResponses(nomisQuestions: List<NomisQuestion>) {
+    this.questions.retainAll(
+      nomisQuestions.map { nomisQuestion ->
+        val question = updateOrAddQuestion(nomisQuestion)
+        question.updateResponses(nomisQuestion.answers)
+        question
+      }.toSet(),
+    )
+  }
 
-    history.clear()
-    addNomisHistory(upsert.history)
+  fun updateHistory(nomisHistory: Collection<NomisHistory>) {
+    nomisHistory.forEach { history ->
```

Comment:
i'm a little confused because different sorted sets of entities are handled differently in the methods here.

should this not have a `this.history.retainAll(…)`? i get that it's a sorted set so if date and type match, a new history item _ought_ not be created, but what about the historical question/response children?

Reply:
This was fixed